### PR TITLE
Fixes #14158 - Add support for tailoring files

### PR DESCRIPTION
**Pull request by xprazak2**
TODO:

- [x] tests

### CODE REVIEW

**Comment by ares** [comment]
trying to update name of tailoring file results in 500

```
2016-11-22T14:37:56 26c05376 [app] [I] Started PATCH "/compliance/tailoring_files/1" for ::1 at 2016-11-22 14:37:56 +0100
2016-11-22T14:37:56 26c05376 [app] [I] Processing by TailoringFilesController#update as HTML
2016-11-22T14:37:56 26c05376 [app] [I]   Parameters: {"utf8"=>"✓", "authenticity_token"=>"D1uNodKYx/YcaE5LYINGIvMwNFi2TigegZ+iW4UOzM1+9PToKnvAgtTCGQjpte6Cnli56y/0NVcw907tdoazgg==", "tailoring_file"=>{"name"=>"my_tailor2", "location_ids"=>["", "2"], "organization_ids"=>["", "1"]}, "commit"=>"Submit", "id"=>"1"}
2016-11-22T14:37:56 26c05376 [sql] [D]   ActiveRecord::SessionStore::Session Load (0.3ms)  SELECT  "sessions".* FROM "sessions" WHERE "sessions"."session_id" = $1  ORDER BY "sessions"."id" ASC LIMIT 1  [["session_id", "26c05376c1e8353aad4a78ba13461875"]]
2016-11-22T14:37:56 26c05376 [sql] [D]   User Load (0.2ms)  SELECT  "users".* FROM "users" WHERE "users"."id" = $1 LIMIT 1  [["id", 3]]
2016-11-22T14:37:56 26c05376 [app] [D] Setting current user thread-local variable to admin
2016-11-22T14:37:56 26c05376 [sql] [D]    (0.3ms)  SELECT COUNT(*) FROM "taxonomies" WHERE "taxonomies"."type" IN ('Organization')
2016-11-22T14:37:56 26c05376 [app] [D] Setting current organization thread-local variable to none
2016-11-22T14:37:56 26c05376 [sql] [D]    (0.2ms)  SELECT COUNT(*) FROM "taxonomies" WHERE "taxonomies"."type" IN ('Location')
2016-11-22T14:37:56 26c05376 [app] [D] Setting current location thread-local variable to none
2016-11-22T14:37:56 26c05376 [sql] [D]   AuthSource Load (0.1ms)  SELECT  "auth_sources".* FROM "auth_sources" WHERE "auth_sources"."id" = $1 LIMIT 1  [["id", 1]]
2016-11-22T14:37:56 26c05376 [sql] [D]    (0.2ms)  SELECT  "taxonomies"."id" FROM "taxonomies" WHERE "taxonomies"."type" IN ('Location') LIMIT 1
2016-11-22T14:37:56 26c05376 [sql] [D]    (0.1ms)  SELECT  "taxonomies"."id" FROM "taxonomies" WHERE "taxonomies"."type" IN ('Organization') LIMIT 1
2016-11-22T14:37:56 26c05376 [sql] [D]   ForemanOpenscap::TailoringFile Load (0.2ms)  SELECT  "foreman_openscap_tailoring_files".* FROM "foreman_openscap_tailoring_files" WHERE "foreman_openscap_tailoring_files"."id" = $1 LIMIT 1  [["id", 1]]
2016-11-22T14:37:56 26c05376 [sql] [D]   ForemanOpenscap::TailoringFile Load (0.3ms)  SELECT "foreman_openscap_tailoring_files".* FROM "foreman_openscap_tailoring_files"
2016-11-22T14:37:56 26c05376 [app] [I] Completed 500 Internal Server Error in 19ms (ActiveRecord: 1.5ms)
2016-11-22T14:37:56 26c05376 [app] [F] 
 | NameError (undefined local variable or method `scap_content_params' for #<TailoringFilesController:0x0055e42e8f7a10>):
 |   /home/ares/Projekty/Zdrojaky/foreman_openscap/app/controllers/tailoring_files_controller.rb:34:in `update'
 |   app/controllers/concerns/application_shared.rb:14:in `set_timezone'
 |   app/models/concerns/foreman/thread_session.rb:32:in `clear_thread'
 |   lib/middleware/catch_json_parse_errors.rb:8:in `call'
 |   lib/middleware/tagged_logging.rb:18:in `call'
 | 
 | 
2016-11-22T14:37:57 26c05376 [app] [I]   Rendered /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/actionpack-4.2.7.1/lib/action_dispatch/middleware/templates/rescues/_source.erb (4.7ms)
2016-11-22T14:37:57 26c05376 [app] [I]   Rendered /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/actionpack-4.2.7.1/lib/action_dispatch/middleware/templates/rescues/_trace.html.erb (1.9ms)
2016-11-22T14:37:57 26c05376 [app] [I]   Rendered /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/actionpack-4.2.7.1/lib/action_dispatch/middleware/templates/rescues/_request_and_response.html.erb (0.7ms)
2016-11-22T14:37:57 26c05376 [app] [I]   Rendered /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/actionpack-4.2.7.1/lib/action_dispatch/middleware/templates/rescues/diagnostics.html.erb within rescues/layout (20.0ms
```

**Comment by ares** [comment]
One more issue found during testing, when I change tailoring file for a policy, the tailoring file is already cached on smart proxy at `tailoring/5/5_tailoring_file.xml`. Even if foreman_scap_client gets new hash, it downloads old tailoring file. We either need to change the way we store files on proxy and start using hashes there too or proxy needs invalidating the files which should be triggered from Foreman whenever tailoring file was changed for policy or the content of the file changed itself. Do we have the same issue for regular scap content file?

Except for this issue it works pretty well :+1:

File: app/models/concerns/foreman_openscap/data_stream_content.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+      validates_with ForemanOpenscap::DataStreamValidator
+
+      after_save :create_profiles
+
+      before_validation :redigest, :if => lambda { |ds_content| ds_content.persisted? && ds_content.scap_file_changed? }
+      before_destroy ActiveRecord::Base::EnsureNotUsedBy.new(:policies)
+    end
+
+    def proxy_url
+      @proxy_url ||= SmartProxy.with_features('Openscap').find do |proxy|
```
Comment (xprazak2) [inline app/models/concerns/foreman_openscap/data_stream_content.rb]: Bad dog

**Comment by xprazak2** [comment]
I added more tests, fixed style most of style issues, disabled a couple of obnoxious cops, fixed the 500. Not sure if we have the problems when changing scap contents for policy, I will have to verify.

File: lib/foreman_openscap/engine.rb [hunk 2]
```diff
@@ -90,6 +90,9 @@ class Engine < ::Rails::Engine
                      :resource_type => 'ForemanOpenscap::ScapContent'
           permission :edit_hosts, { :hosts => [:openscap_proxy_changed] }, :resource_type => "Host"
           permission :edit_hostgroups, { :hostgroups => [:openscap_proxy_changed] }, :resource_type => "Hostgroup"
+          permission :view_tailoring_files, { :tailoring_files => [:index] }, :resource_type => 'ForemanOpenscap::TailoringFile'
```
Comment (ares) [inline lib/foreman_openscap/engine.rb]: should be probably added to roles "Compliance viewer" and "Compliance manager" (others to manager role)

File: app/views/tailoring_files/welcome.html.erb [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  <h1><%= _('Tailoring Files') %></h1>
+  <p><%= _('It may sometimes be required to adjust the security policy to your specific needs. ') %></br>
+    <%= (_('In Foreman, tailoring_files represent the custom modifications to default XCCDF profiles and they can be applied to hosts
+    via %s') % link_to('compliance policies', policies_path)).html_safe %>
+  </p>
+
+  <div class="blank-slate-pf-main-action">
+    <%= new_link(_('New Tailoring File'), {}, { :class => "btn-lg" }) %>
+  </div>
+</div>
```
Comment (ares) [inline app/views/tailoring_files/welcome.html.erb]: empty line missing

Comment (xprazak2) [inline app/views/tailoring_files/welcome.html.erb]: There would be no-newline character if it was missing.

File: app/controllers/policies_controller.rb [hunk 4]
```diff
@@ -4,16 +4,16 @@ class PoliciesController < ApplicationController
...
     ::ForemanOpenscap::Policy
   end
 
   def index
-    @policies = resource_base
-                  .search_for(params[:search], :order => params[:order])
-                  .paginate(:page => params[:page], :per_page => params[:per_page])
-                  .includes(:scap_content, :scap_content_profile)
+    @policies = resource_base.search_for(params[:search], :order => params[:order])
+                             .paginate(:page => params[:page], :per_page => params[:per_page])
```
Comment (ares) [inline app/controllers/policies_controller.rb]: +1

File: app/controllers/concerns/foreman/controller/parameters/tailoring_file.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
+module Foreman::Controller::Parameters::TailoringFile
```
Comment (ares) [inline app/controllers/concerns/foreman/controller/parameters/tailoring_file.rb]: I'd disable this cop

File: app/helpers/policies_helper.rb [hunk 6]
```diff
@@ -38,6 +46,26 @@ def scap_content_profile_selector(form)
     end
   end
 
+  def tailoring_file_selector(form)
+    select_f form, :tailoring_file_id, ForemanOpenscap::TailoringFile.all, :id, :name,
```
Comment (ares) [inline app/helpers/policies_helper.rb]: `ForemanOpenscap::TailoringFile.all` missing `.authorized(:view_tailoring_files)`?

File: app/views/tailoring_files/welcome.html.erb (hunk 3 above)
Comment (ares) [inline app/views/tailoring_files/welcome.html.erb]: not missing anymore, I saw the red sign before :-) pls dismiss

**Comment by ares** [comment]
Mostly looks good, I left few comments inline. I need to do more testing. It would also be great if we added a check for openscap module version on proxy and would only use proxies with the right version. Any ideas how it could behave from UI and implementation perspective? Maybe it would be enough to just don't enable tailoring file upload unless there's smart proxy openscap with version X and explaining why on index page? Last but not least, we should add API coverage I think.

**Comment by xprazak2** [comment]
Disabling the tailoring file upload if there is no proxy with openscap version X would work not entirely work if there are proxies with different openscap versions. You would be able to upload a tailoring file (because all you need is an arbitrary proxy with version X) but policy with tailoring file could be applied to a host that has proxy with old version W.

The upload part is easy, it can be even handled in the validator and we do not have to touch UI at all.
The rest is a bit more tricky, all that I could think of is a 'Version health check' button of some sort. It would have to find all existing hosts with policy where policy has a tailoring file and openscap proxy of the host has version lower than X. 

We definitely will want to API (and hammer commands) for this as well, I was planning it separately, but I can add it here if you'd like.

**Comment by xprazak2** [comment]
Changes:

* I moved the dots to end of lines and enabled the corresponding cop
* Tailoring file XML can be downloaded from UI through action menu
* When uploading a tailoring file with old proxy version, notification is shown
* grouping dropdowns for tailoring file and its profiles in policy wizard, tailoring profiles dropdown is not shown when tailoring file is not selected

**Comment by xprazak2** [comment]
Adjusted column width on tailoring_files/index page.

**Comment by ares** [comment]
Cool, you even added Created column :+1: I think it's ready for merge now pending jenkins. Once we get smart proxy PR also confirmed by Jenkins I'll merge the whole train of PRs

**Comment by shlomizadok** [comment]
Thanks @xprazak2, @ares :tada:

**Comment by ares** [comment]
:+1: @xprazak2 I think you have a good material for next community demo :-)

**Pull request merged**